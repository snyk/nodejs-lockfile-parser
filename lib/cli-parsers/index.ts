import { DepGraph, DepGraphBuilder } from '@snyk/dep-graph';
import { getYarnLockfileType } from '..';
import { LockfileType } from '../parsers';
import { FormattedCliOutput } from './cli-parser-types';
import { extractNameAndIdentifier } from './cli-parser-utils';
import { parseYarnInfoOutput } from './yarn-info-parser';
import { parseYarnListOutput } from './yarn-list-parser';

export const buildDepGraphFromCliOutput = (
  rawCliOutput: string,
  lockfileContent: string,
  manifestFileContent: string,
): DepGraph => {
  const manifestDependencies: Record<string, string> =
    JSON.parse(manifestFileContent).dependencies || {};

  const lockfileType = getYarnLockfileType(lockfileContent);

  const { name: rootName, version: rootVersion } =
    JSON.parse(manifestFileContent);

  const pkgManagerVersion: '1' | '2' =
    lockfileType === LockfileType.yarn ? '1' : '2';

  // Build depMap object from the cli output
  const formattedCliOutput: FormattedCliOutput =
    pkgManagerVersion === '1'
      ? parseYarnListOutput(rawCliOutput, manifestDependencies)
      : parseYarnInfoOutput(rawCliOutput);

  const rootPkgInfo: { name: string; version?: string } = rootName
    ? { name: rootName, ...(rootVersion && { version: rootVersion }) }
    : undefined;

  const pkgManager = {
    name: 'yarn',
    version: pkgManagerVersion,
  };

  const builder = new DepGraphBuilder(pkgManager, rootPkgInfo);

  const { topLevelDeps, dependencies: depMap } = formattedCliOutput;

  // Add all nodes
  [...depMap.keys()].forEach((name) => {
    const { name: pkgName, identifier: pkgVersion } =
      extractNameAndIdentifier(name);
    builder.addPkgNode(
      { name: pkgName, version: pkgVersion.split(':').pop() as string },
      name,
    );
  });

  // Deal with root special case first
  const rootNodeId = builder.rootNodeId;
  topLevelDeps.forEach((dep) => builder.connectDep(rootNodeId, dep));

  // Now rest of deps
  [...depMap.entries()].forEach(([parent, deps]) => {
    deps.forEach((dep) => {
      builder.connectDep(parent, dep);
    });
  });

  return builder.build();
};
