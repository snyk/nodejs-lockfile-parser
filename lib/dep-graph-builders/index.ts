import {
  parseYarnLockV1Project,
  parseYarnLockV1WorkspaceProject,
  buildDepGraphYarnLockV1SimpleCyclesPruned,
  buildDepGraphYarnLockV1Simple,
  buildDepGraphYarnLockV1WorkspaceCyclesPruned,
  buildDepGraphYarnLockV1Workspace,
  extractPkgsFromYarnLockV1,
} from './yarn-lock-v1';
import {
  buildDepGraphYarnLockV2Simple,
  parseYarnLockV2Project,
  extractPkgsFromYarnLockV2,
} from './yarn-lock-v2';
import { parseNpmLockV2Project } from './npm-lock-v2';

export {
  parseNpmLockV2Project,
  parseYarnLockV1Project,
  buildDepGraphYarnLockV1Workspace,
  buildDepGraphYarnLockV1SimpleCyclesPruned,
  buildDepGraphYarnLockV1Simple,
  buildDepGraphYarnLockV1WorkspaceCyclesPruned,
  parseYarnLockV1WorkspaceProject,
  extractPkgsFromYarnLockV1,
  buildDepGraphYarnLockV2Simple,
  parseYarnLockV2Project,
  extractPkgsFromYarnLockV2,
};
