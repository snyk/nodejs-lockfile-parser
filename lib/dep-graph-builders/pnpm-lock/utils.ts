import { OutOfSyncError } from '../../errors';
import { LockfileType } from '../../parsers';
import { getChildNode, PkgNode } from '../util';
import { PnpmNormalisedPkgs } from './type';

export const getChildNodePnpmLockV7Workspace = (
  name: string,
  depInfo: { version: string; isDev: boolean },
  workspacePkgNameToVersion: Record<string, string>,
  pkgs: PnpmNormalisedPkgs,
  strictOutOfSync: boolean,
  includeOptionalDeps: boolean,
) => {
  let childNode: PkgNode;

  if (workspacePkgNameToVersion[name]) {
    const version = workspacePkgNameToVersion[name];

    // This is just to mimic old behavior where when StrictOutOfSync is turned on,
    // any cross referencing between workspace packages will lead to a throw
    if (strictOutOfSync) {
      throw new OutOfSyncError(`${name}@${version}`, LockfileType.pnpm);
    }

    childNode = {
      id: `${name}@${version}`,
      name: name,
      version: version,
      dependencies: {},
      isDev: depInfo.isDev,
    };
  } else {
    childNode = getChildNode(
      name,
      depInfo,
      pkgs,
      strictOutOfSync,
      includeOptionalDeps,
    );
  }

  return childNode;
};
