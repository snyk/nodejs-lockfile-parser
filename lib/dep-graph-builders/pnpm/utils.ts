import { LockfileType } from '../..';
import { getGraphDependencies } from '../util';
import { PnpmLockfileParser } from './lockfile-parser/lockfile-parser';
import { NormalisedPnpmPkgs, PnpmNode } from './types';
import { valid } from 'semver';
import { OpenSourceEcosystems } from '@snyk/error-catalog-nodejs-public';
import {
  INSTALL_COMMAND,
  LOCK_FILE_NAME,
} from '../../errors/out-of-sync-error';

export const getPnpmChildNode = (
  name: string,
  depInfo: { version: string; isDev: boolean },
  pkgs: NormalisedPnpmPkgs,
  strictOutOfSync: boolean,
  includeOptionalDeps: boolean,
  lockfileParser: PnpmLockfileParser,
): PnpmNode => {
  const resolvedVersion =
    valid(depInfo.version) || depInfo.version === undefined
      ? depInfo.version
      : lockfileParser.excludeTransPeerDepsVersions(depInfo.version);
  const childNodeKey = `${name}@${resolvedVersion}`;
  if (!pkgs[childNodeKey]) {
    if (lockfileParser.isWorkspaceLockfile()) {
      return {
        id: childNodeKey,
        name: name,
        version: resolvedVersion,
        dependencies: {},
        isDev: depInfo.isDev,
      };
    }
    if (strictOutOfSync && !/^file:/.test(depInfo.version)) {
      const errMessage =
        `Dependency ${childNodeKey} was not found in ` +
        `${LOCK_FILE_NAME[LockfileType.pnpm]}. Your package.json and ` +
        `${
          LOCK_FILE_NAME[LockfileType.pnpm]
        } are probably out of sync. Please run ` +
        `"${INSTALL_COMMAND[LockfileType.pnpm]}" and try again.`;
      throw new OpenSourceEcosystems.PnpmOutOfSyncError(errMessage);
    } else {
      return {
        id: childNodeKey,
        name: name,
        version: resolvedVersion,
        dependencies: {},
        isDev: depInfo.isDev,
        missingLockFileEntry: true,
      };
    }
  } else {
    const depData = pkgs[childNodeKey];
    const dependencies = getGraphDependencies(
      depData.dependencies || {},
      depInfo.isDev,
    );
    const optionalDependencies = includeOptionalDeps
      ? getGraphDependencies(depData.optionalDependencies || {}, depInfo.isDev)
      : {};
    return {
      id: `${name}@${depData.version}`,
      name: name,
      version: resolvedVersion,
      dependencies: { ...dependencies, ...optionalDependencies },
      isDev: depInfo.isDev,
    };
  }
};
