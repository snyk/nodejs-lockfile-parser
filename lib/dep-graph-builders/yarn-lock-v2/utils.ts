import { structUtils } from '@yarnpkg/core';
import * as _flatMap from 'lodash.flatmap';
import { OutOfSyncError } from '../../errors';
import { LockfileType } from '../../parsers';
import { NormalisedPkgs } from '../types';
import { getGraphDependencies, PkgNode } from '../util';

const BUILTIN_PLACEHOLDER = 'builtin';
const MULTIPLE_KEYS_REGEXP = / *, */g;

export type ParseDescriptor = typeof structUtils.parseDescriptor;
export type ParseRange = typeof structUtils.parseRange;

const keyNormalizer =
  (parseDescriptor: ParseDescriptor, parseRange: ParseRange) =>
  (rawDescriptor: string): string[] => {
    // See https://yarnpkg.com/features/protocols
    const descriptors: string[] = [rawDescriptor];
    const descriptor = parseDescriptor(rawDescriptor);
    const name = `${descriptor.scope ? '@' + descriptor.scope + '/' : ''}${
      descriptor.name
    }`;
    const range = parseRange(descriptor.range);
    const protocol = range.protocol;
    switch (protocol) {
      case 'npm:':
      case 'file:':
        // This is space inneficient but will be kept for now,
        // Due to how we wish to index using the dependencies map
        // we want the keys to match name@version but this is handled different
        // for npm alias and normal install.
        descriptors.push(`${name}@${range.selector}`);
        descriptors.push(`${name}@${protocol}${range.selector}`);
        break;
      case 'git:':
      case 'git+ssh:':
      case 'git+http:':
      case 'git+https:':
      case 'github:':
        if (range.source) {
          descriptors.push(
            `${name}@${protocol}${range.source}${
              range.selector ? '#' + range.selector : ''
            }`,
          );
        } else {
          descriptors.push(`${name}@${protocol}${range.selector}`);
        }
        break;
      case 'patch:':
        if (range.source && range.selector.indexOf(BUILTIN_PLACEHOLDER) === 0) {
          descriptors.push(range.source);
        } else {
          descriptors.push(
            `${name}@${protocol}${range.source}${
              range.selector ? '#' + range.selector : ''
            }`,
          );
        }
        break;
      case null:
      case undefined:
        if (range.source) {
          descriptors.push(`${name}@${range.source}#${range.selector}`);
        } else {
          descriptors.push(`${name}@${range.selector}`);
        }
        break;
      case 'http:':
      case 'https:':
      case 'link:':
      case 'portal:':
      case 'exec:':
      case 'workspace:':
      case 'virtual:':
      default:
        // For user defined plugins
        descriptors.push(`${name}@${protocol}${range.selector}`);
        break;
    }
    return descriptors;
  };

export type YarnLockFileKeyNormalizer = (fullDescriptor: string) => Set<string>;

export const yarnLockFileKeyNormalizer =
  (
    parseDescriptor: ParseDescriptor,
    parseRange: ParseRange,
  ): YarnLockFileKeyNormalizer =>
  (fullDescriptor: string) => {
    const allKeys = fullDescriptor
      .split(MULTIPLE_KEYS_REGEXP)
      .map(keyNormalizer(parseDescriptor, parseRange));
    return new Set<string>(_flatMap(allKeys));
  };

export const getYarnLockV2ChildNode = (
  name: string,
  depInfo: { version: string; isDev: boolean },
  pkgs: NormalisedPkgs,
  strictOutOfSync: boolean,
  includeOptionalDeps: boolean,
  resolutions: Record<string, string>,
  parentNode: PkgNode,
) => {
  // First check if a resolution would be used
  const resolvedVersionFromResolution = (() => {
    // First check for scoped then simple
    const scopedKey = `${parentNode.name}/${name}`;
    if (resolutions[scopedKey]) {
      return resolutions[scopedKey];
    } else if (resolutions[name]) {
      return resolutions[name];
    }
    return '';
  })();

  if (resolvedVersionFromResolution) {
    const childNodeKeyFromResolution = `${name}@${resolvedVersionFromResolution}`;
    const {
      version: versionFromResolution,
      dependencies,
      optionalDependencies,
    } = pkgs[childNodeKeyFromResolution];

    const formattedDependencies = getGraphDependencies(
      dependencies || {},
      depInfo.isDev,
    );
    const formattedOptionalDependencies = includeOptionalDeps
      ? getGraphDependencies(optionalDependencies || {}, depInfo.isDev)
      : {};

    return {
      id: childNodeKeyFromResolution,
      name: name,
      version: versionFromResolution,
      dependencies: {
        ...formattedOptionalDependencies,
        ...formattedDependencies,
      },
      isDev: depInfo.isDev,
    };
  }

  // No resolutions
  const childNodeKey = `${name}@${depInfo.version}`;
  if (!pkgs[childNodeKey]) {
    if (strictOutOfSync && !/^file:/.test(depInfo.version)) {
      throw new OutOfSyncError(childNodeKey, LockfileType.yarn2);
    } else {
      return {
        id: childNodeKey,
        name: name,
        version: depInfo.version,
        dependencies: {},
        isDev: depInfo.isDev,
        missingLockFileEntry: true,
      };
    }
  } else {
    const depData = pkgs[childNodeKey];
    const dependencies = getGraphDependencies(
      depData.dependencies || {},
      depInfo.isDev,
    );
    const optionalDependencies = includeOptionalDeps
      ? getGraphDependencies(depData.optionalDependencies || {}, depInfo.isDev)
      : {};
    return {
      id: `${name}@${depData.version}`,
      name: name,
      version: depData.version,
      dependencies: { ...dependencies, ...optionalDependencies },
      isDev: depInfo.isDev,
    };
  }
};
