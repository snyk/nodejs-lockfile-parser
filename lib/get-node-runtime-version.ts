export default function getRuntimeVersion(): number {
  return parseInt(process.version.slice(1).split('.')[0], 10);
}
