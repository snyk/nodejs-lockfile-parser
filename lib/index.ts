import 'source-map-support/register';
import * as fs from 'fs';
import * as path from 'path';
import * as _ from 'lodash';

export {
  buildDepTree,
  buildDepTreeFromFiles,
};

interface PkgTree {
  name: string;
  version: string;
  dependencies?: {
    [dep: string]: PkgTree;
  };
}

async function buildDepTree(targetFileRaw: string, lockFileRaw: string): Promise<PkgTree> {

  const lockFile = JSON.parse(lockFileRaw);
  const targetFile = JSON.parse(targetFileRaw);

  if (!targetFile.dependencies) {
    throw new Error("No 'dependencies' property in package.json");
  }
  if (!lockFile.dependencies) {
    throw new Error("No 'dependencies' property in package-lock.json");
  }

  const depTree: PkgTree = {
    dependencies: {},
    name: targetFile.name,
    version: targetFile.version,
  };

  const topLevelDeps = Object.keys(targetFile.dependencies);

  await Promise.all(topLevelDeps.map(async (dep) => {
    depTree.dependencies[dep] = await buildSubTreeRecursive(dep, [], lockFile);
  }));

  return depTree;
}

async function buildSubTreeRecursive(dep: string, depKeys: string[], lockFile: object): Promise<PkgTree> {

  const depSubTree: PkgTree = {
    dependencies: {},
    name: dep,
    version: undefined,
  };

  // Get path to the nested dependencies from list ['package1', 'package2']
  // to ['dependencies', 'package1', 'dependencies', 'package2', 'dependencies']
  const depPath = getDepPath(depKeys);
  // try to get list of deps on the path
  const deps = _.get(lockFile, depPath);

  // If exists and looked-up dep is there
  if (deps && deps[dep]) {
    // update the tree
    depSubTree.version = deps[dep].version;
    // repeat the process for dependencies of looked-up dep
    const newDeps = deps[dep].requires ? Object.keys(deps[dep].requires) : [];
    await Promise.all(newDeps.map(async (subDep) => {
      depSubTree.dependencies[subDep] = await buildSubTreeRecursive(subDep, [...depKeys, dep], lockFile);
    }));
    return depSubTree;
  } else {
    // tree was walked to the root and dependency was not found
    if (!depKeys.length) {
      throw new Error(`Dependency ${dep} was not found in package-lock.json.
        Your package.json and package-lock.json are probably out of sync.
        Please run npm install and try to parse the log again.`);
    }
    // dependency was not found on a current path, remove last key (move closer to the root) and try again
    return buildSubTreeRecursive(dep, depKeys.slice(0, -1), lockFile);
  }
}

function getDepPath(depKeys: string[]) {
  const depPath = depKeys.reduce((acc, key) => {
        return acc.concat([key, 'dependencies']);
      }, ['dependencies']);

  return depPath;
}

async function buildDepTreeFromFiles(root: string, targetFilePath: string, lockFilePath: string): Promise<PkgTree> {
  if (!root || !lockFilePath || !lockFilePath) {
    throw new Error('Missing required parameters for parseLockFile()');
  }

  const targetFileFullPath = path.resolve(root, targetFilePath);
  const lockFileFullPath = path.resolve(root, lockFilePath);

  const asyncExists = promisify(fs.stat);
  try {
    await asyncExists(targetFileFullPath);
  } catch (e) {
    throw new Error(`Target file package.json not found at location: ${targetFileFullPath}`);
  }

  try {
    await asyncExists(lockFileFullPath);
  } catch (e) {
    throw new Error(`LockFile package-lock.json not found at location: ${lockFileFullPath}`);
  }

  const asyncReadFile = promisify(fs.readFile);
  const targetFile = await asyncReadFile(targetFileFullPath, 'utf-8');
  const lockFile = await asyncReadFile(lockFileFullPath, 'utf-8');

  return await buildDepTree(targetFile, lockFile);
}

function promisify(func): any {
  return (...args: any[]) => {
    return new Promise((resolve, reject) => {
      return func(...args, (err, data) => {
        if (err) {
          return reject(err);
        }
        resolve(data);
      });
    });
  };
}
