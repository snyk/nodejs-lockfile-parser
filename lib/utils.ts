import { readFileSync } from 'fs';
import { InvalidUserInputError } from './errors';

export enum NodeLockfileVersion {
  NpmLockV1 = 'NPM_LOCK_V1',
  NpmLockV2 = 'NPM_LOCK_V2',
  NpmLockV3 = 'NPM_LOCK_V3',
  YarnLockV1 = 'YARN_LOCK_V1',
  YarnLockV2 = 'YARN_LOCK_V2',
}

export const getLockfileVersionFromFile = (
  targetFile: string,
): NodeLockfileVersion => {
  const lockFileContents = readFileSync(targetFile, 'utf-8');
  if (targetFile.endsWith('package-lock.json')) {
    return getNpmLockfileVersion(lockFileContents);
  } else if (targetFile.endsWith('yarn.lock')) {
    return getYarnLockfileVersion(lockFileContents);
  } else {
    throw new InvalidUserInputError(
      `Unknown lockfile ${targetFile}. ` +
        'Please provide either package-lock.json or yarn.lock.',
    );
  }
};

export function getYarnLockfileVersion(
  lockFileContents: string,
): NodeLockfileVersion.YarnLockV1 | NodeLockfileVersion.YarnLockV2 {
  if (lockFileContents.includes('__metadata')) {
    return NodeLockfileVersion.YarnLockV2;
  } else {
    return NodeLockfileVersion.YarnLockV1;
  }
}

export function getNpmLockfileVersion(
  lockFileContents: string,
):
  | NodeLockfileVersion.NpmLockV1
  | NodeLockfileVersion.NpmLockV2
  | NodeLockfileVersion.NpmLockV3 {
  try {
    const lockfileJson = JSON.parse(lockFileContents);
    const lockfileVersion: number | null = lockfileJson.lockfileVersion || null;

    switch (lockfileVersion) {
      case null:
      case 1:
        return NodeLockfileVersion.NpmLockV1;
      case 2:
        return NodeLockfileVersion.NpmLockV2;
      case 3:
        return NodeLockfileVersion.NpmLockV3;
      default:
        throw new InvalidUserInputError(
          `Unsupported npm lockfile version in package-lock.json. ` +
            'Please provide a package-lock.json with lockfileVersion 1, 2 or 3',
        );
    }
  } catch (e) {
    throw new InvalidUserInputError(
      `Problem parsing package-lock.json - make sure the package-lock.json is a valid JSON file`,
    );
  }
}
