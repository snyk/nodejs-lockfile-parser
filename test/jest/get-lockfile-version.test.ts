import { getLockfileVersionFromFile } from '../../lib/utils';

describe('getLockfileVersionFromFile', () => {
  it('npm', () => {
    const result = getLockfileVersionFromFile(
      `${__dirname}/../fixtures/bare-npm/package-lock.json`,
    );
    expect(result).toMatchSnapshot();
  });
});
