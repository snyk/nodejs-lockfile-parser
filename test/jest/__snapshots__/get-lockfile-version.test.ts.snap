// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLockfileVersionFromFile npm 1`] = `"NPM_LOCK_V2"`;
